ch_backup/storage/engine/s3/s3_retry.py

```diff
@@ -57,3 +61,92 @@ def wrapper(self: "S3StorageEngine", *args: Any, **kwargs: Any) -> RT:
                 raise S3RetryingError(f"Failed to make S3 operation: {str(e)}") from e
 
         return wrapper
+
+
+def retry(
+    exception_types: Union[type, tuple] = Exception,
+    max_attempts: int = 5,
+    max_interval: float = 5,
+    multiplier: float = 0.5,
+    sleep_function: Callable = sleep,
+    verbose: bool = True,
+) -> Callable:
+    """
+    Decorator for thread safe retry logic with exponential wait time.
+    """
+    retry_exponential = RetryExponential(
+        exception_types, max_attempts, max_interval, multiplier, sleep_function, verbose
+    )
+
+    def wrap(f):
+        @wraps(f)
+        def wrapped_f(*args, **kwargs):
+            return retry_exponential(f, *args, **kwargs)
+
+        return wrapped_f
+
+    return wrap
+
+
+class RetryExponential:
+    """
+    Exponential retry.
+    """
+
+    def __init__(
+        self,
+        exception_types: Union[type, tuple] = Exception,
+        max_attempts: int = 5,
+        max_interval: float = 5,
+        multiplier: float = 0.5,
+        sleep_function: Callable = sleep,
+        verbose: bool = True,
+    ) -> None:
+        self.exception_types = exception_types
+        self.max_attempts = max_attempts
+        self.max_interval = max_interval
+        self.multiplier = multiplier
+        self.sleep_function = sleep_function
+        self.verbose = verbose
+
+    def calculate_sleep_time(self, attempt):
+        """
+        Calculate time to sleep.
+        """
+
+        high_value = (2**attempt) * self.multiplier
+        return uniform(0, min(high_value, self.max_interval))  # nosec
```

Comment:
Nevertheless, I think fixed interval backoff is more preferable here. Pay attention on [this](https://github.com/jd/tenacity/blob/310058274ed22a345e9c3917c97b4afd6363d5a5/tenacity/wait.py#L186) suggestion from tenacity. Clickhouse itself uses exponential fixed backoff intervals when requsting S3.